README.md
# Stream Language
Check it out at [Pub.Dev](https://pub.dev/packages/multi_language_json)

A simple way to support your Flutter application with multiple languages!

![Gravar_2020_04_12_05_50_22_436](https://user-images.githubusercontent.com/22732544/79065252-6aa64e80-7c85-11ea-99f4-32904a624331.gif)

## Help Maintenance

I've been maintaining quite many repos these days and burning out slowly. If you could help me cheer up, buying me a cup of coffee will make my life really happy and get much energy out of it.

<a href="https://www.buymeacoffee.com/RtrHv1C" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/purple_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>

## Getting Started
First you need create a directory in your main folder with name lang like this:

![Screenshot_20](https://user-images.githubusercontent.com/22732544/79065270-81e53c00-7c85-11ea-87cb-e39040d1cb6a.png)

Note that each json file must have its language and country code as its name.
Remember to add them to your pubspec:

![Screenshot_21](https://user-images.githubusercontent.com/22732544/79065273-83166900-7c85-11ea-914b-3ae277816c36.png)

In the code, to start you must first create an object with the following attributes:
    final language = MultiLanguageBloc(
          initialLanguage: 'pt_BR',
          defaultLanguage: 'pt_BR',
          commonRoute: 'common',
          supportedLanguages: [
            'en_US', 'pt_BR'
          ]
      );

MultiLanguageBloc is a singleton, after the first start, it will have the same attributes.

### CommonRoute:
Here you enter the key within the language that contains words that can be used on more than one screen as in the example below:

![Screenshot_22](https://user-images.githubusercontent.com/22732544/79065283-94f80c00-7c85-11ea-9d39-05b6e38bdcc3.png)

The first time you use firebase language you should do this:

    final language = MultiLanguageBloc(
       initialLanguage: 'pt_BR',
       defaultLanguage: 'pt_BR',
       commonRoute: 'common',
       supportedLanguages: [
         'en_US', 'pt_BR'
       ]
    );

    @override
    Widget build(BuildContext context) {
      return MultiLanguageStart( // # Use this widget to init languages, this will only be necessary once, after that you will only use the 'MultiStreamLanguage' widget
        future: language.init(), // # Here you call the method to init
        builder: (c) => MultiStreamLanguage(
          screenRoute: ['home'],
          builder: (context, data) => Scaffold(
            appBar: AppBar(
              title: Text(d.getValue(route: ['title'])),
            ),
            body: Center(
              child: RaisedButton(
                child: Text(data.getValue(route: ['btn'])),
                onPressed: () => language.showAlertChangeLanguage(
                  context: context,
                  title: data.getValue(
                    route: ['dialog', 'title'],
                    inRoute: false
                  ),
                  btnNegative: data.getValue(
                      route: ['dialog', 'btn_negative'],
                      inRoute: false
                  )
                ),
              ),
            ),
          )
        )
      );
    }

From the next you start using only the `MultiStreamLanguage` widget, the first one is needed because the first app should load all language and start the default language from the user's mobile language.

## Widget MultiStreamLanguage

### ScreenRoute
This is where the magic happens, as a parameter it receives the screen route within the language key, see that in the code above is as:
`screenRoute: ['home']`, in json it looks like this:

![Screenshot_23](https://user-images.githubusercontent.com/22732544/79065294-a17c6480-7c85-11ea-9e89-e34f30fdd2fe.png)

If the route were a node within 'home' you would go something like this: `screenRoute: ['home', 'route_inside_home']`

### [Very Important] Builder -> Data
Here is the parent object to access the texts, for that use the getValue method that has **route** as parameter, here it will work like ScreenRoute, its second parameter is **inRoute**, if true, the passed route will be traveled inside the path gives wheel informed in **ScreenRoute**, if false, it will look in the parent route, that is, in the beginning of json, ex:

if i called `data.getValue(route: ['my_route'], inRoute: true/false);`

![Screenshot_23](https://user-images.githubusercontent.com/22732544/79065294-a17c6480-7c85-11ea-9e89-e34f30fdd2fe.png)

#### with inRoute true:
It will return: **B**

#### with inRoute false:
It will return: **A**

Another very interesting point, if the past route does not exist in the current language, it will search in the standard language, and if it does not exist there, it will return a string with the text 'NULL', ex:

![Gravar_2020_04_12_05_49_28_636~1](https://user-images.githubusercontent.com/22732544/79065457-ce7d4700-7c86-11ea-8da2-21a03b78f62c.gif)

# Changing Language
For this, every language node must have a child named config with the following attributes:

![Screenshot_18](https://user-images.githubusercontent.com/22732544/79065332-e7d1c380-7c85-11ea-9380-8262ba6b5a8d.png)

The **title** key is what will appear as the language name in the dialog you select.

After that you can call the method:

    language.showAlertChangeLanguage(
        context: context,
        title: 'title',
        btnNegative: 'text'
    )

This will show an alert dialog like this (Language and flag listing is done automatically from the data passed in the **config** node):

![Screenshot_19](https://user-images.githubusercontent.com/22732544/79065375-4139f280-7c86-11ea-94bc-7202dfad3be1.png)

To change the language programmatically, just call this method passing as the language prefix ex:

    language.changeLanguage('pt_BR');


CHANGELOG.md
## [1.2.0] - Null Safety

## [1.1.0+2] - Format

## [1.1.0+1] - Fix dependencies

## [1.1.0] - Fix
* Fix bug in getter "currentValue"

## [1.0.0+2] - Fix dependencies.

## [1.0.0+1] - Fixed warnings.

## [1.0.0] - TODO: Add release date.

* TODO: Describe initial release.

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:multi_language_json/multi_language_json.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Multi Language - Example',
        home: Home(),
        supportedLocales: [const Locale('en', 'US'), const Locale('pt', 'BR')]);
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final language = MultiLanguageBloc(
      initialLanguage: 'pt_BR',
      defaultLanguage: 'pt_BR',
      commonRoute: 'common',
      supportedLanguages: ['en_US', 'pt_BR']);

  @override
  Widget build(BuildContext context) {
    return MultiLanguageStart(
        future: language.init(),
        builder: (c) => MultiStreamLanguage(
            screenRoute: ['home'],
            builder: (c, d) => Scaffold(
                  appBar: AppBar(title: Text(d.getValue(route: ['title']))),
                  body: Center(
                    child: ElevatedButton(
                      child: Text(d.getValue(route: ['btn'])),
                      onPressed: () => language.showAlertChangeLanguage(
                          context: context,
                          title: d.getValue(
                              route: ['dialog', 'title'], inRoute: false),
                          btnNegative: d.getValue(
                              route: ['dialog', 'btn_negative'],
                              inRoute: false)),
                    ),
                  ),
                )));
  }
}


lib/multi_language.dart
library multi_language;

import 'dart:convert';
import 'package:flutter/services.dart' show rootBundle;
import 'package:bloc_pattern/bloc_pattern.dart';
import 'package:country_pickers/country.dart';
import 'package:country_pickers/country_pickers.dart';
import 'package:devicelocale/devicelocale.dart';
import 'package:flutter/material.dart';
import 'package:rxdart/rxdart.dart';

class MultiStreamLanguage extends StatelessWidget {
  final Function onChange;
  final List<String> screenRoute;
  final languageBloc = MultiLanguageBloc._instance;
  final Widget Function(BuildContext c, LangSupport data) builder;

  MultiStreamLanguage({@required this.builder, this.screenRoute = const [], this.onChange}){
    if(onChange != null){
      languageBloc.outStreamList.listen((v) => onChange());
    }
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder(
      initialData: languageBloc.currentValue,
      stream: languageBloc.outStreamList,
      builder: (context, s) {
        dynamic globalScreenRoute = languageBloc
            ._languages[languageBloc.defaultLanguage];
        dynamic screenRoute = s.data;

        this.screenRoute.forEach((v){
          if(globalScreenRoute != null) globalScreenRoute = globalScreenRoute[v];
          if(screenRoute != null) screenRoute = screenRoute[v];
        });

        return builder(context, LangSupport(
            languageBloc._languages[languageBloc.defaultLanguage],
            s.data,
            globalScreenRoute,
            screenRoute,
            languageBloc.commonRoute
        ));
      },
    );
  }
}

class MultiLanguageBloc implements BlocBase {
  String lastLanguage;
  final _languages = {};
  final String commonRoute;
  final String defaultLanguage;
  final List<String> supportedLanguages;
  final _language = BehaviorSubject<Map<String, dynamic>>();

  Stream<dynamic> get outStreamList => _language.stream;
  Map<dynamic, dynamic> get currentValue => _language.value;
  Map<dynamic, dynamic> get currentCommon =>
      commonRoute != null ? currentValue[commonRoute] : [];

  static MultiLanguageBloc _instance;
  factory MultiLanguageBloc({@required List<String> supportedLanguages, @required String defaultLanguage, @required String initialLanguage, String commonRoute}) {
    _instance ??= MultiLanguageBloc._internal(
        supportedLanguages: supportedLanguages,
        defaultLanguage: defaultLanguage,
        lastLanguage: initialLanguage,
        commonRoute: commonRoute);
    return _instance;
  }

  MultiLanguageBloc._internal(
      {this.supportedLanguages, this.defaultLanguage, this.lastLanguage, this.commonRoute});

  Future<void> init() async {
    this.lastLanguage ??= defaultLanguage;
    for(int i = 0; i < supportedLanguages.length; i++){
      this._languages[supportedLanguages[i]] = await parseJsonFromAssets('lang/' + supportedLanguages[i]);
    }

    await _setDeviceLanguage();
  }

  Future<MultiLanguageBloc> _setDeviceLanguage() async {
    await changeLanguage(await Devicelocale.currentLocale);
    return this;
  }

  Future<void> changeLanguage(String prefix) async {
    if (_languages[prefix] == null) {
      prefix = this.defaultLanguage;
      print(
          'setting language with defaultLanguage because prefix: $prefix dont exists in map!');
    }

    if (this.lastLanguage != prefix) {
      this.lastLanguage = prefix;
      _language.sink.add(_languages[this.lastLanguage]);
      print('language inited with language: $prefix');
    } else {
      print(
          'language dont changed because informed prefix is the same as current language: $prefix');
    }
  }

  List<Map<dynamic, dynamic>> getListLanguage() {
    return _languages.values.map<Map>((v) => v['config']).toList();
  }

  Future<dynamic> showAlertChangeLanguage(
      {@required context,
        @required String title,
        @required String btnNegative}) async {
    List<Map<dynamic, dynamic>> out = this.getListLanguage();

    return await showDialog(
        context: context,
        builder: (context) => AlertDialog(
          title: Text(title),
          actions: <Widget>[
            FlatButton(
              child: Text(btnNegative),
              onPressed: () => Navigator.pop(context),
            ),
          ],
          content: Container(
            width: MediaQuery.of(context).size.height * 0.8,
            height: MediaQuery.of(context).size.height * 0.3,
            child: ListView.builder(
              itemCount: out.length,
              itemBuilder: (BuildContext context, int index) {
                return Material(
                  color: currentValue['config']['prefix'] ==
                      out[index]['prefix']
                      ? Colors.blueAccent[700]
                      : Colors.transparent,
                  child: ListTile(
                    leading: CountryPickerUtils.getDefaultFlagImage(
                        Country(isoCode: out[index]['iso_code'])),
                    selected: currentValue['config']['prefix'] ==
                        out[index]['prefix'],
                    title: Text(
                      out[index]['title'].toString(),
                      style: TextStyle(
                          color: currentValue['config']['prefix'] ==
                              out[index]['prefix']
                              ? Colors.white
                              : Colors.black),
                    ),
                    onTap: () {
                      changeLanguage(out[index]['prefix']);
                      Navigator.pop(context);
                    },
                  ),
                );
              },
            ),
          ),
        ));
  }

  Future<Map<String, dynamic>> parseJsonFromAssets(String assetsPath) async {
    return rootBundle.loadString(assetsPath + '.json')
        .then((jsonStr) => jsonDecode(jsonStr));
  }

  @override
  void dispose() {
    _language.close();
  }

  @override
  void addListener(listener) {}

  @override
  bool get hasListeners => null;

  @override
  void notifyListeners() {}

  @override
  void removeListener(listener) {}
}

class MultiLanguageStart extends StatelessWidget {
  final future;
  final Widget loadWidget;
  final Function(BuildContext context) builder;
  MultiLanguageStart(
      {@required this.future, @required this.builder, this.loadWidget});

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: future,
      builder: (context, snapshot) {
        if (snapshot.connectionState == ConnectionState.done) {
          return builder(context);
        }
        return loadWidget ?? CircularProgressIndicator();
      },
    );
  }
}

class LangSupport{
  final _defaultLang;
  final _currentLang;
  final _defaultRouteLang;
  final _currentRouteLang;
  final _commonKey;

  LangSupport(this._defaultLang,
      this._currentLang,
      this._defaultRouteLang,
      this._currentRouteLang,
      this._commonKey
      );

  dynamic getCommon(){
    return getValue(
        inRoute: false,
        route: [this._commonKey]
    );
  }

  dynamic getValue({@required List<String> route, bool inRoute = true}){
    dynamic toReturn = inRoute ? _currentRouteLang : _currentLang;
    dynamic toHelp = inRoute ? _defaultRouteLang : _defaultLang;

    route.forEach((e) {
      if(toHelp != null) toHelp = toHelp[e];
      if(toReturn != null) toReturn = toReturn[e];
    });

    return toReturn ?? (toHelp ?? 'NULL');
  }
}

lib/multi_language_json.dart
library multi_language_json;

import 'package:country_pickers/country.dart';
import 'package:flutter/services.dart' show rootBundle;
import 'package:country_pickers/country_pickers.dart';
import 'package:flutter/material.dart';
import 'package:rxdart/rxdart.dart';
import 'dart:convert';

class MultiStreamLanguage extends StatelessWidget {
  /// On language change
  final Function? onChange;

  /// route in json, ex:
  /// {
  ///   "a" : {
  ///     "b" : "value"
  ///   }
  /// }
  ///
  /// to get only "b" I pass: ['a', 'b']
  final List<String> screenRoute;
  final languageBloc = MultiLanguageBloc._instance;

  /// Widget child
  final Widget Function(BuildContext c, LangSupport data) builder;

  MultiStreamLanguage(
      {required this.builder, this.screenRoute = const [], this.onChange}) {
    if (onChange != null) {
      languageBloc?.outStreamList.listen((v) => onChange!());
    }
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder(
        initialData: languageBloc?.currentValue,
        stream: languageBloc?.outStreamList,
        builder: (context, s) {
          dynamic globalScreenRoute =
              languageBloc?._languages[languageBloc?.defaultLanguage];
          dynamic screenRoute = s.data;

          this.screenRoute.forEach((v) {
            if (globalScreenRoute != null)
              globalScreenRoute = globalScreenRoute[v];
            if (screenRoute != null) screenRoute = screenRoute[v];
          });

          return builder(
              context,
              LangSupport(
                  languageBloc?._languages[languageBloc?.defaultLanguage],
                  s.data,
                  globalScreenRoute,
                  screenRoute,
                  languageBloc?.commonRoute));
        });
  }
}

class MultiLanguageBloc implements _Bloc {
  /// Last language selected
  String lastLanguage;

  /// List ofs languages
  final _languages = {};

  /// Route common in to screens
  final String? commonRoute;

  /// Default language
  final String defaultLanguage;

  /// List of supported languages ex: ['en_US', 'pt_BR']
  /// Here you pass the names of files in folder "lang" without the ".json"
  final List<String> supportedLanguages;
  final _language = BehaviorSubject<Map<String, dynamic>>();

  /// Stream of languages
  Stream<dynamic> get outStreamList => _language.stream;

  /// Current json [Full]
  Map<dynamic, dynamic> get currentValue => _languages[lastLanguage];

  /// Current common in current json selected
  Map<dynamic, dynamic> get currentCommon =>
      commonRoute != null ? currentValue[commonRoute] : [];

  static MultiLanguageBloc? _instance;
  factory MultiLanguageBloc(
      {required List<String> supportedLanguages,
      required String defaultLanguage,
      String? initialLanguage,
      String? commonRoute}) {
    return _instance ??= MultiLanguageBloc._internal(
        supportedLanguages: supportedLanguages,
        defaultLanguage: defaultLanguage,
        lastLanguage: initialLanguage ?? defaultLanguage,
        commonRoute: commonRoute);
  }

  MultiLanguageBloc._internal(
      {required this.supportedLanguages,
      required this.defaultLanguage,
      required this.lastLanguage,
      this.commonRoute});

  /// Required called before all to load all jsons
  Future<void> init() async {
    for (int i = 0; i < supportedLanguages.length; i++) {
      this._languages[supportedLanguages[i]] =
          await parseJsonFromAssets('lang/' + supportedLanguages[i]);
    }

    await _setDeviceLanguage();
  }

  Future<MultiLanguageBloc> _setDeviceLanguage() async {
    await changeLanguage(this.defaultLanguage);
    return this;
  }

  /// Change current language passing prefix ex: [changeLanguage('en_US')]
  Future<void> changeLanguage(String prefix) async {
    if (_languages[prefix] == null) {
      prefix = this.defaultLanguage;
      print(
          'setting language with defaultLanguage because prefix: $prefix dont exists in map!');
    }

    if (this.lastLanguage != prefix) {
      this.lastLanguage = prefix;
      _language.sink.add(_languages[this.lastLanguage]);
      print('language inited with language: $prefix');
    } else {
      print(
          'language dont changed because informed prefix is the same as current language: $prefix');
    }
  }

  /// Get list of languages in route "config" inside jsons
  List<Map<dynamic, dynamic>> getListLanguage() {
    return _languages.values.map<Map>((v) => v['config']).toList();
  }

  /// Show alert to change language
  Future<dynamic> showAlertChangeLanguage(
      {required BuildContext context,
      required String title,
      required String btnNegative}) async {
    List<Map<dynamic, dynamic>> out = this.getListLanguage();

    return await showDialog(
        context: context,
        builder: (context) => AlertDialog(
              title: Text(title),
              actions: <Widget>[
                TextButton(
                    child: Text(btnNegative),
                    onPressed: () => Navigator.pop(context)),
              ],
              content: Container(
                width: MediaQuery.of(context).size.height * 0.8,
                height: MediaQuery.of(context).size.height * 0.3,
                child: ListView.builder(
                  itemCount: out.length,
                  itemBuilder: (BuildContext context, int index) {
                    return Material(
                      color: currentValue['config']['prefix'] ==
                              out[index]['prefix']
                          ? Colors.blueAccent[700]
                          : Colors.transparent,
                      child: ListTile(
                        leading: CountryPickerUtils.getDefaultFlagImage(Country(
                            isoCode: out[index]['iso_code'],
                            iso3Code: '',
                            name: '',
                            phoneCode: '')),
                        selected: currentValue['config']['prefix'] ==
                            out[index]['prefix'],
                        title: Text(
                          out[index]['title'].toString(),
                          style: TextStyle(
                              color: currentValue['config']['prefix'] ==
                                      out[index]['prefix']
                                  ? Colors.white
                                  : Colors.black),
                        ),
                        onTap: () {
                          changeLanguage(out[index]['prefix']);
                          Navigator.pop(context);
                        },
                      ),
                    );
                  },
                ),
              ),
            ));
  }

  Future<Map<String, dynamic>> parseJsonFromAssets(String assetsPath) async {
    return rootBundle
        .loadString(assetsPath + '.json')
        .then((jsonStr) => jsonDecode(jsonStr));
  }

  @override
  void dispose() {
    _language.close();
  }
}

class MultiLanguageStart extends StatelessWidget {
  final future;
  final Widget? loadWidget;
  final Function(BuildContext context) builder;

  MultiLanguageStart(
      {required this.future, required this.builder, this.loadWidget});

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: future,
      builder: (context, snapshot) {
        if (snapshot.connectionState == ConnectionState.done) {
          return builder(context);
        }
        return loadWidget ?? CircularProgressIndicator();
      },
    );
  }
}

class LangSupport {
  final _defaultLang;
  final _currentLang;
  final _defaultRouteLang;
  final _currentRouteLang;
  final _commonKey;

  LangSupport(this._defaultLang, this._currentLang, this._defaultRouteLang,
      this._currentRouteLang, this._commonKey);

  dynamic getCommon() {
    return getValue(inRoute: false, route: [this._commonKey]);
  }

  dynamic getValue({required List<String> route, bool inRoute = true}) {
    dynamic toReturn = inRoute ? _currentRouteLang : _currentLang;
    dynamic toHelp = inRoute ? _defaultRouteLang : _defaultLang;

    route.forEach((e) {
      if (toHelp != null) toHelp = toHelp[e];
      if (toReturn != null) toReturn = toReturn[e];
    });

    return toReturn ?? (toHelp ?? 'NULL');
  }
}

abstract class _Bloc {
  void dispose();
}
